### PR TITLE
feat: eme error interface

### PR DESCRIPTION
**Pull request by adrums86**
This adds new error constants from video.js for a more cohesive errors interface, this also adds some new events for granular insight into the EME workflow.

### CODE REVIEW

File: src/plugin.js [hunk 1]
```diff
@@ -188,9 +188,12 @@ export const setupSessions = (player) => {
  */
 export const emeErrorHandler = (player) => {
   return (objOrErr) => {
+    const vjsErrorType = videojs.Error && videojs.Error.EMEKeySessionCreationError;
+    const errorType = vjsErrorType || 'eme-key-session-creation-error';
```
Comment (dzianis-dashkevich) [inline src/plugin.js:192]: I guess we should not fallback.
We should expect `videojs.Error.EMEKeySessionCreationError` is available and we should update minimum videojs peerDependency version

**Comment by dzianis-dashkevich** [comment]
Do we plan to include other errors?

**Comment by adrums86** [comment]
I added an error constant for each keysystem event and added videojs as a peerDependency. Changing this PR to a draft until the constants merge into video.js

**Comment by adrums86** [comment]
Can't flip this back to a draft PR for whatever reason. This shouldn't be merged until the corresponding video.js chore is merged and released first: https://github.com/videojs/video.js/pull/8634

File: src/eme.js [hunk 2]
```diff
@@ -82,10 +82,14 @@ export const getSupportedKeySystem = (keySystems) => {
...
+        window.navigator.requestMediaKeySystemAccess(keySystem, supportedConfigurations).catch((error) => {
+          emeError(error, videojs.Error.EMEFailedToRequestMediaKeySystemAccess);
+        });
     } else {
-      promise = promise.catch((e) =>
-        window.navigator.requestMediaKeySystemAccess(keySystem, supportedConfigurations));
+      promise = promise.catch(() => {
+        // TODO: Not sure if we want to call emeError here or not.
+        window.navigator.requestMediaKeySystemAccess(keySystem, supportedConfigurations);
+      });
```
Comment (dzianis-dashkevich) [inline src/eme.js:92]: Here is just an accumulation of the promise,
I guess you should trigger error once in the other place

CleanShot 2024-03-12 at 13 12 11@2x

File: src/plugin.js [hunk 3]
```diff
@@ -241,8 +250,10 @@ const onPlayerReady = (player, emeError) => {
     player.tech_.el_.addEventListener('encrypted', (event) => {
       videojs.log.debug('eme', 'Received an \'encrypted\' event');
       setupSessions(player);
-      handleEncryptedEvent(player, event, getOptions(player), player.eme.sessions, player.tech_)
-        .catch(emeError);
+      handleEncryptedEvent(player, event, getOptions(player), player.eme.sessions, player.tech_, emeError)
+        .catch((error) => {
+          emeError(error, videojs.Error.EMEEncryptedError);
```
Comment (dzianis-dashkevich) [inline src/plugin.js:255]: what is `videojs.Error.EMEEncryptedError`?

File: src/plugin.js [hunk 4]
```diff
@@ -258,13 +269,17 @@ const onPlayerReady = (player, emeError) => {
       try {
         handleMsNeedKeyEvent(event, getOptions(player), player.eme.sessions, player.tech_);
       } catch (error) {
-        emeError(error);
+        emeError(error, videojs.Error.MSKeyError);
       }
     });
-    player.tech_.on('mskeyerror', emeError);
+    const msKeyErrorCallback = (error) => {
+      emeError(error, videojs.Error.MSKeyError);
```
Comment (dzianis-dashkevich) [inline src/plugin.js:276]: Why do we have key system-specific errors?
I think we should use common interface for all of them

File: src/plugin.js [hunk 5]
```diff
@@ -362,7 +377,9 @@ const eme = function(options = {}) {
         // element between sources
         setupSessions(player);
         handleWebKitNeedKeyEvent(event, getOptions(player), player.tech_)
-          .catch(emeError);
+          .catch((error) => {
+            emeError(error, videojs.Error.WebkitKeyError);
+          });
```
Comment (dzianis-dashkevich) [inline src/plugin.js:382]: as we discussed offline:
we can analyze each rejection here, and try to use spec errors for legacy flow

File: src/plugin.js [hunk 6]
```diff
@@ -334,7 +349,7 @@ const eme = function(options = {}) {
           if (event.type === 'mskeyerror') {
             callback(event.target.error);
             if (!suppressErrorIfPossible) {
-              emeError(event.message);
+              emeError(event.message, videojs.Error.MSKeyError);
```
Comment (dzianis-dashkevich) [inline src/plugin.js:352]: as we discussed offline:
we can analyze each rejection here, and try to use spec errors for legacy flow

File: src/plugin.js [hunk 7]
```diff
@@ -319,12 +334,12 @@ const eme = function(options = {}) {
...
 
       if (window.MediaKeys) {
-        handleEncryptedEvent(player, mockEncryptedEvent, mergedEmeOptions, player.eme.sessions, player.tech_)
+        handleEncryptedEvent(player, mockEncryptedEvent, mergedEmeOptions, player.eme.sessions, player.tech_, emeError)
           .then(() => callback())
           .catch((error) => {
             callback(error);
             if (!suppressErrorIfPossible) {
-              emeError(error);
+              emeError(error, videojs.Error.EMEEncryptedError);
```
Comment (dzianis-dashkevich) [inline src/plugin.js:342]: as discussed offline:
we should probably analyze possible rejections here to narrow it to specific error or remove it

File: src/plugin.js (hunk 4 above)
Comment (dzianis-dashkevich) [reply src/plugin.js:276]: discussed offline

File: src/plugin.js (hunk 3 above)
Comment (dzianis-dashkevich) [reply src/plugin.js:255]: discussed offline

**Review by dzianis-dashkevich — APPROVED** [review]

**Pull request merged**